src/drivers/i2c/adapters/imx_i2c/imx_i2c.c

```diff
@@ -0,0 +1,266 @@
+	delay(1);
+	tmp &= ~(IMX_I2C_I2CR_MSTA) ;
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2CR, tmp);
+	delay(10);
+	tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2CR);
+	tmp &= ~(IMX_I2C_I2CR_IEN) ;
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2CR, tmp);
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2CR, 0);
+
+	return 0;
+}
+
+static int imx_i2c_start(struct imx_i2c *adapter) {
+	uint32_t tmp;
+
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2SR, 0 );
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2CR, IMX_I2C_I2CR_IEN );
+
+	delay(100);
+
+	/* Start I2C transaction */
+	tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2CR);
+	tmp |= IMX_I2C_I2CR_MSTA ;
+	REG8_STORE(adapter->base_addr+ IMX_I2C_I2CR, tmp);
+
+	tmp = imx_i2c_bus_busy(adapter, 1);
+	if (tmp) {
+		return -1;
+	}
+
+	tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2CR);
+	tmp |= IMX_I2C_I2CR_MTX | IMX_I2C_I2CR_TXAK;
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2CR, tmp);
+
+	return 0;
+}
+#if 0
+static int imx_i2c_tx_byte(struct imx_i2c *adapter, uint8_t byte) {
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2SR, 0);
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2DR, byte);
+	res = imx_i2c_trx_complete(adapter);
+	if (res) {
+		goto out;
+	}
+
+}
+#endif
+
+static int imx_i2c_rx(struct imx_i2c *adapter, uint16_t addr, uint8_t *buff, size_t sz) {
+	int res = -1;
+	int cnt;
+	uint32_t tmp;
+
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2SR, 0);
+	/* write slave address */
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2DR, ((addr << 1) | 0x1) & 0xFF);
+	res = imx_i2c_trx_complete(adapter);
+	if (res) {
+		goto out;
+	}
+	if (REG8_LOAD(adapter->base_addr + IMX_I2C_I2SR) &  IMX_I2C_I2SR_RXAK) {
+		res = -ENODEV;
+		goto out;
+	}
+
+	log_debug("ACK received 0x%x", addr);
+	tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2CR);
+	tmp &= ~IMX_I2C_I2CR_MTX;
+	if (sz > 1) {
+		tmp &= ~IMX_I2C_I2CR_TXAK;
+	}
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2CR, tmp);
+
+	REG8_STORE(adapter->base_addr + IMX_I2C_I2SR, 0);
+	/* dummy read */
+	tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2DR);
+
+	res = sz;
+
+	for (cnt = sz; cnt > 0; cnt--) {
+		tmp = imx_i2c_trx_complete(adapter);
+		if (tmp) {
+			log_debug("i2c complition error");
+			res = -1;
+			goto out;
+		}
+
+		log_debug("i2c complition ok");
+		res = 1;
+
+		if (cnt == 1) {
+			/*
+			 * It must generate STOP before read I2DR to prevent
+			 * controller from generating another clock cycle
+			 */
+			tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2CR);
+			tmp &= ~(IMX_I2C_I2CR_MTX | IMX_I2C_I2CR_MSTA);
+			REG8_STORE(adapter->base_addr + IMX_I2C_I2CR, tmp);
+		}
+		if (cnt == 2) {
+			tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2CR);
+			tmp |= IMX_I2C_I2CR_TXAK;
+			REG8_STORE(adapter->base_addr + IMX_I2C_I2CR, tmp);
+		}
+
+		imx_i2c_bus_busy(adapter, 0);
+		REG8_STORE(adapter->base_addr + IMX_I2C_I2SR, 0);
+		tmp = REG8_LOAD(adapter->base_addr + IMX_I2C_I2DR);
+
+		log_debug("read 0x%x ", tmp);
+		*buff++ = (uint8_t)(tmp & 0xFF);
```

Comment:
`tmp` is used as 8-bit integer in this function, may be just declare it as `uint8_t`?